### ISSUE TITLE
How to fix CS9236?

### ISSUE THREAD

**Issue by Evangelink** [dotnet/roslyn#78347]
I have 3 diagnostics CS9236 and every change I did around explicit type is still triggering the issue.

```csharp
int totalTokens = llmInfo.Sum(x => x.Value.Sum(y => y.Item1 + y.Item2));
int totalInputToken = llmInfo.Sum(x => x.Value.Sum(y => y.Item1));
int totalOutputToken = llmInfo.Sum(x => x.Value.Sum(y => y.Item2));
```

with `llmInfo` of type `KeyValuePair<string, List<(int, int)>>`.

**Comment by CyrusNajmabadi** [comment]
If llmInfo is a KVP, then it won't have .Sum defined on it. You need to do llmInfo.Value.Sum(...)

**Comment by Youssef1313** [comment]
@CyrusNajmabadi The code here compiles. This is about CS9236 which was done to warn for deeply nested lambda resolution, but the case above feels simple and it's unclear why it does so much lambda binding, and why it doesn't go away with specifying explicit type. Please re-open.

Ping @cston as well as you implemented that in #72823

**Comment by Evangelink** [comment]
Sorry I meant to write `List<KeyValuePair<string, List<(int, int)>>>`. I am not blocked as I switched over to a foreach but it's the 3rd time I am unable to fix such diagnostic so I would guess more people would also struggle.

**Comment by CyrusNajmabadi** [comment]
What happens if you write these as explicitly typed lambdas? (Types for parameters and for the return type of the lambda)?

**Comment by jjonescz** [comment]
Looks like adding an explicit return type helps, e.g.,:

```cs
int totalTokens = llmInfo.Sum(x => x.Value.Sum(int (y) => y.Item1 + y.Item2));
```

**Comment by jaredpar** [comment]
The CS9236 diagnostic is an informational message from the compiler that a particular lambda considered an unusually high number of overloads. This means that the compilation is spending significant amount of time binding that lambda. These types of issues can be the root of long compilation times. For example: we've seen situations where compile times of a project were in the 2m range because of 1-2 lambdas in this form. 

The best way to address this issue, as @jjonescz noted, is to explicitly type the lambda. That removes the need for the compiler to run type inference which is usually the root of the high number of overloads considered.
